### PR TITLE
Include additional constrained MO problems

### PR DESCRIPTION
**Pull request by cyrilpic**
I have implemented two additional constrained test suite for pymoo that could be of interest to others.

I am not so sure about what to put in the documentation. Let me know.

### CODE REVIEW

**Comment by julesy89** [comment]
Awesome! Thank you very much for your contribution!
I would be more than happy to see more :) 

The implementation of test problems can be really tricky and it is easy to make mistakes. I am speaking of experience. One thing I always try to avoid is making mistakes in test problems because they have an impact on researcher's results.

**My workflow to ensure correctness is the following:**

What I always do with test problems is to pull out the original implementation published by the authors. I have done that for DASCMOP (http://imagelab.stu.edu.cn/Content.aspx?type=content&Content_ID=1310, link on the bottom, implemented in JAVA).

Then, I usually run their code for random inputs and store the results in files. In my last commit you can see the files here: 
https://github.com/msu-coinlab/pymoo/tree/master/tests/problems/resources/DASCOMP
Created with difficulties of `0.25, 0.0, 0.0`.

Afterwards, I create test cases to match this input:
https://github.com/msu-coinlab/pymoo/blob/master/tests/problems/test_problems_dascomp.py

For DASCMOP5, I had to set the variable number to `5` to match the results. What is suggested by the authors in the paper? 

DASCMOP8 and DASCMOP9 do not pass the tests yet. Can you please check if there is a reason for that? 
I sometimes compare the code published by the authors with mine and then with the paper. It can be tricky sometimes to find the source of truth though.

I am really looking forward it to include it into the framework soon!

**Comment by cyrilpic** [comment]
I have fully tested the MW test suite against the original C++ implementation. I can include this in the tests.

I had checked DAS-CMOP against the Java implementation, but well let me double check (the number of variables is always 30 according to the original publication).

**Comment by julesy89** [comment]
Perfect! The test cases do have the purpose to show correctness and to avoid changing something in the problems accidentally later on.

I usually prefer the code published by the authors. However, if the source of the C++ code is reliable that works too! Feel free to replace the test files with yours of you are confident they are correct. Please document in the test file what implementation you have used to create them.
Maybe it makes also sense to use two or three different difficulties as well.

For WFG I had two different source codes and their implementation was slightly different. Then, one has to decide what is "more" correct. This might be the case here too.

Can you please do the same for the MW test problems?
Thanks!

**Comment by julesy89** [comment]
The test problems passes the tests this way:

```
class DASCMOP8(DASCMOP7):
    def _evaluate(self, X, out, *args, **kwargs):
        g = self.g2(X)
        f0 = np.cos(0.5 * np.pi * X[:, 0:1]) * np.cos(0.5 * np.pi * X[:, 1:2]) + g
        f1 = np.cos(0.5 * np.pi * X[:, 0:1]) * np.sin(0.5 * np.pi * X[:, 1:2]) + g
        f2 = np.sin(0.5 * np.pi * X[:, 0:1]) + g
        out["F"] = np.column_stack([f0, f1, f2])
        out["G"] = self.constraints(X, f0, f1, f2, g)

```

**Comment by julesy89** [comment]
For the DAS-CMOP 9 the g3 was problematic: 

 ```
   def g3(self, X):
        j = np.arange(self.n_obj - 1, self.n_var) + 1
        contrib = (X[:, self.n_obj - 1:] - np.cos(0.25 * j / self.n_var * np.pi * (X[:, 0:1] + X[:, 1:2]))) ** 2
        return contrib.sum(axis=1)[:, None]
```

There needs to be a `j = np.arange(self.n_obj - 1, self.n_var) + 1` with the plus one because the paper assumes an index 1 based notation. I will commit my changes. Please have a look and validate the correctness.

**Comment by cyrilpic** [comment]
That was fast. Nice catch for the errors in DAS-CMOP, thanks. Your changes look good to me.
As it seems you can generate tests very quickly from JAVA, I think it would be wise to check another difficulty triplet such as `[0.5, 1.0, 0.5]`.

I'll right away open a new pull request for the MW test cases.

**Comment by julesy89** [comment]
I was about to do that anyway. Please see my most recent commit;
Right now for difficulty `[0.0, 1.0, 0.0]` there is a deviation to the JAVA code.
See the new tests for `dasmop`.

Can you please check what is going wrong there?

I think when the tests are passing now, the implementation is ready to be added to the framework.

**Comment by cyrilpic** [comment]
Yes I will double check that. I saw there was a difference between the formula in the publication and the implementation.

File: pymoo/problems/multi/dascmop.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        q_k = np.array([[1.5, 0.5, 2.5, 1.5, 0.5, 3.5, 2.5, 1.5, 0.5]])
+
+        a_k2 = 0.3
+        b_k2 = 1.2
+        theta_k = -0.25*np.pi
+
+        c = np.zeros((X.shape[0], 2+p_k.shape[1]))
+
+        c[:, 0] = np.sin(a*np.pi*X[:, 0]) - b
+        c[:, 1:2] = (e - g)*(g - d)
```
Comment (cyrilpic) [inline pymoo/problems/multi/dascmop.py:63]:
```python
        if self.zeta == 1.:
            c[:, 1:2] = 1e-4 - np.abs(e - g)
        else:
            c[:, 1:2] = (e - g) * (g - d)
```

File: pymoo/problems/multi/dascmop.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+        x_k = np.array([[1.0, 0., 0., 1.0 / np.sqrt(3.0)]])
+        y_k = np.array([[0., 1.0, 0., 1.0 / np.sqrt(3.0)]])
+        z_k = np.array([[0., 0., 1.0, 1.0 / np.sqrt(3.0)]])
+
+        c = np.zeros((X.shape[0], 3+x_k.shape[1]))
+
+        c[:, 0] = np.sin(a*np.pi*X[:, 0]) - b
+        c[:, 1] = np.cos(a*np.pi*X[:, 1]) - b
+        c[:, 2:3] = (e - g)*(g - d)
```
Comment (cyrilpic) [inline pymoo/problems/multi/dascmop.py:157]:
```python 
       if self.zeta == 1:
            c[:, 2:3] = 1e-4 - np.abs(e - g)
        else:
            c[:, 2:3] = (e - g) * (g - d)
```

**Comment by cyrilpic** [comment]
In the JAVA code, they do a distinction if zeta == 1. I commented the code with the corrections. It's probably the easiest if you push the changes directly.

**Comment by julesy89** [comment]
Thanks! I fixed it and all the tests are passing now.
(I am going to commit soon).

Do you have data files with the Pareto fronts? 
Not sure if I am doing it right, but the one from the Java implementation seem to be totally different than what I expect.

**Comment by cyrilpic** [comment]
[dascmop.zip](https://github.com/msu-coinlab/pymoo/files/4595255/dascmop.zip)

I found this somewhere on the authors' website.

**Comment by julesy89** [comment]
Nice. I have added them. But just to be sure.
For every difficulty vector there is a different Pareto front?
The file structure looks like that...

**Comment by julesy89** [comment]
And probably the same question for the `MW` suite. 

For the test problems so far, most people use them only when the fronts are somehow available. I have gotten some requests and questions about that in the past.

**Comment by cyrilpic** [comment]
Yes each difficulty vector is linked to a different Pareto front.

Here the fronts for MW: 
[MW.zip](https://github.com/msu-coinlab/pymoo/files/4597839/MW.zip)

The calculation of the Pareto fronts for MW1, MW2, MW3, MW4, MW6, MW8 and MW12 is already implemented in the committed python. For the others it's slightly less trivial.

**Comment by julesy89** [comment]
Perfect. I have added now the missing fronts and they are loaded from files. I personally like to have the equations implemented but whenever this becomes too complicated it is sometimes not worth the effort.

The problem implementations are now done I think. The last step to make them being used by users is to explain them and their purpose. We do not need to provide any equations just some properties and visualizations are sufficient. 

**Can you please provide users some more information about the test problems?**
I have already created two files:
doc/source/problem/constrained/mw.ipynb 
doc/source/problem/constrained/dascmopipynb 

If have already linked them on the index page and on the problem index page.

Moreover, I would like to mention your name or alias on the bottom of both test problems to give credit to your effort of implementing them! (Only if that is okay for you)

**Pull request merged**